File: MicroTDCore/Sources/MicroTDCore/Entities/Tower.swift
// Tower.swift
// Runtime tower entity

import Foundation

/// Runtime tower instance placed on grid
public final class Tower {
    public let instanceId: Int  // Deterministic ID for rendering
    public let typeID: String
    public let position: GridPosition
    public let baseDef: TowerDef
    
    // Runtime state
    public private(set) var ticksSinceLastFire: Int = 0
    public private(set) var kills: Int = 0
    public private(set) var totalDamage: Double = 0
    
    // Modified stats (from relics/upgrades)
    public var damageMultiplier: Double = 1.0
    public var rangeMultiplier: Double = 1.0
    public var fireRateMultiplier: Double = 1.0
    public var slowOnHit: Double = 0.0
    
    public init(instanceId: Int, typeID: String, position: GridPosition, baseDef: TowerDef) {
        self.instanceId = instanceId
        self.typeID = typeID
        self.position = position
        self.baseDef = baseDef
    }
    
    public var effectiveDamage: Double {
        return baseDef.damage * damageMultiplier
    }
    
    public var effectiveRange: Double {
        return baseDef.range * rangeMultiplier
    }
    
    public var effectiveFireRate: Double {
        return baseDef.fireRate * fireRateMultiplier
    }
    
    public var ticksPerShot: Int {
        return Int(Double(SimulationClock.ticksPerSecond) / effectiveFireRate)
    }
    
    public func canFire() -> Bool {
        return ticksSinceLastFire >= ticksPerShot
    }
    
    public func incrementTick() {
        ticksSinceLastFire += 1
    }
    
    public func resetFireCooldown() {
        ticksSinceLastFire = 0
    }
    
    public func recordDamage(_ damage: Double) {
        totalDamage += damage
    }
    
    public func recordKill() {
        kills += 1
    }
}

File: expo/src/core/entities/Tower.ts
// Tower.ts
// Runtime tower entity

import type { TowerDef } from '../definitions/TowerDef';
import { GridPosition } from './GridPosition';
import { SimulationClock } from '../SimulationClock';

/**
 * Runtime tower instance placed on grid
 */
export class Tower {
    // Runtime state
    private _ticksSinceLastFire = 0;
    private _kills = 0;
    private _totalDamage = 0;

    // Modified stats (from relics/upgrades)
    damageMultiplier = 1.0;
    rangeMultiplier = 1.0;
    fireRateMultiplier = 1.0;
    slowOnHit = 0.0;

    constructor(
        public readonly instanceId: number,  // Deterministic ID for rendering
        public readonly typeID: string,
        public readonly position: GridPosition,
        public readonly baseDef: TowerDef
    ) {
        this.slowOnHit = this.baseDef.slowOnHit || 0.0;
    }

    get ticksSinceLastFire(): number {
        return this._ticksSinceLastFire;
    }

    get kills(): number {
        return this._kills;
    }

    get totalDamage(): number {
        return this._totalDamage;
    }

    get effectiveDamage(): number {
        return this.baseDef.damage * this.damageMultiplier;
    }

    get effectiveRange(): number {
        return this.baseDef.range * this.rangeMultiplier;
    }

    get effectiveFireRate(): number {
        return this.baseDef.fireRate * this.fireRateMultiplier;
    }

    get ticksPerShot(): number {
        return Math.floor(SimulationClock.TICKS_PER_SECOND / this.effectiveFireRate);
    }

    canFire(): boolean {
        return this._ticksSinceLastFire >= this.ticksPerShot;
    }

    incrementTick(): void {
        this._ticksSinceLastFire++;
    }

    resetFireCooldown(): void {
        this._ticksSinceLastFire = 0;
    }

    recordDamage(damage: number): void {
        this._totalDamage += damage;
    }

    recordKill(): void {
        this._kills++;
    }
}
